ROSETTA1-166 A Rust canister that transfers funds using ICP Ledger API

MarioDfinity opened the pull request:
Overview
This PR adds an simple example of Ledger usage. The Tokens Transfer canister allows to check the canister balance and transfer tokens from its account to other accounts.

roman-kashitsyn commented on rust/tokens_transfer/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+    to_subaccount: Option<Subaccount>,
+}
+
+// returns the memo as transaction id
```
I think we should return the block index as well. I'm not sure why would we return memo...

MarioDfinity replied:
Removed the memo and returned BlockIndex

roman-kashitsyn commented on rust/tokens_transfer/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, Hash)]
+pub struct TransferArgs {
+    amount: Tokens,
+    to_account: Principal,
```
```suggestion
    to_principal: Principal,
```
I find it confusing: we identify accounts with 32-byte arrays, not principals.

MarioDfinity replied:
The idea is to also show how to create account id from principals.

MarioDfinity replied:
I applied the change with the last commit

roman-kashitsyn commented on rust/tokens_transfer/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+    Ok(memo)
+}
+
+fn hash_transfer_input(args: &TransferArgs, conf: &Conf) -> u64 {
+    let mut hasher = DefaultHasher::default();
+    std::time::SystemTime::now().hash(&mut hasher);
+    id().hash(&mut hasher);
+    args.hash(&mut hasher);
+    conf.hash(&mut hasher);
+    hasher.finish()
+}
```
I'm not sure I understand the purpose of constructing the memo this way.
We could use a fixed memo or at least provide an explanation why we do it this way, otherwise people might assume that this is the way you should always construct memos.

MarioDfinity replied:
Removed.

roman-kashitsyn commented on rust/tokens_transfer/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, Hash, PartialEq)]
+pub struct Conf {
+    ledger_canister_id: Principal,
+    subaccount: Option<Subaccount>,
```
Whose subaccount it is? Subaccount that this canister should always use?
I find it a bit weird: why would you fix a single subaccount? If you only allow one, just use the default one.

MarioDfinity replied:
I think it makes sense in the context of an example to show that there are several subaccounts for each account.

MarioDfinity replied:
I pushed a comment and add a small description and a link to the Ledger doc.

roman-kashitsyn commented on rust/tokens_transfer/src/errors.rs:
```diff
@@ -0,0 +1,29 @@
+use ic_types::PrincipalError;
+use serde::{Deserialize, Serialize};
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub enum TransferError {
+    AccountError(String),
+    CallError(String),
+    LedgerError(String),
+}
+
+impl From<PrincipalError> for TransferError {
+    fn from(e: PrincipalError) -> Self {
+        TransferError::AccountError(format!("{}", e))
+    }
+}
+
+impl From<(RejectionCode, String)> for TransferError {
+    fn from(e: (RejectionCode, String)) -> Self {
+        TransferError::CallError(format!("rejection_code:{:?} error:{}", e.0, e.1))
+    }
+}
+
+impl From<ic_ledger_types::TransferError> for TransferError {
+    fn from(e: ic_ledger_types::TransferError) -> Self {
+        TransferError::LedgerError(format!("{}", e))
+    }
+}
```
Not sure it makes sense to have an elaborate error data type in the example, given that this canister is not designed as an API for other canisters. Returning an error as a string is fine IMO.

MarioDfinity replied:
Removed and returning a string.

roman-kashitsyn commented on rust/tokens_transfer/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+    Ok(res.0?)
+}
+
+#[test]
+fn check_candid_interface_compatibility() {
+    candid::export_service!();
+    let actual_interface = __export_service();
+    let mut actual = tempfile::NamedTempFile::new().expect("Failed to create a temporary file");
+    write!(actual, "{}", actual_interface).unwrap();
+    let expected = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("src").join("tokens_transfer.did");
+    candid_util::check_candid_interface_compatibility(&expected, &actual.path());
+}
```
Not sure we need to track backward compatibility here: this check is _really_ important for heavily-used production canisters, but for examples it's just unnecessary complexity.

roman-kashitsyn commented on rust/tokens_transfer/README.md:
```diff
@@ -0,0 +1,71 @@
+# of the minting principal and your principal respectively
+read -r -d '' ARGS <<EOM
+(record {
+     minting_account="${MINTING_PRINCIPAL_HEX}";
+     initial_values=vec { record { "${YOUR_PRINCIPAL_HEX}"; record { e8s=10_000_000_000 } }; };
```
Are you sure those are _principals_ and  not _accounts ids_?

MarioDfinity replied:
Changed to account ids.

roman-kashitsyn commented on rust/tokens_transfer/README.md:
```diff
@@ -0,0 +1,71 @@
+# TOKENS_TRANSFER_PRINCIPAL_ID_BYTES is the vec nat8 representation of the tokens transfer canister
+read -r -d '' ARGS <<EOM
+(record {
+  to=${TOKENS_TRANSFER_PRINCIPAL_ID_BYTES};
```
I'm pretty sure this is an account id, not principal.

MarioDfinity replied:
Changed to account ids.

roman-kashitsyn commented:
As `ic-ledger-types` has to depend `ic-cdk` anyways (doesn't compile without it for unclear reasons), I decided to add `account_balance` and `transfer` async functions to make the client code a bit nicer: https://github.com/dfinity/cdk-rs/pull/168/commits/83bf346160f4f89cfded9545585dd8eacf0c0b1e
Feel free to use them in your example.

MarioDfinity commented:
I think it's a good idea to expose those functions for client code. I've changed the example to use ic_ledger_types::transfer and removed my utility function. Thank you.

MarioDfinity commented:
@roman-kashitsyn I renamed to_account to to_principal to better describe what the field is and to avoid confusion with AccountIdentifiers.

roman-kashitsyn commented on rust/tokens_transfer/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Test Locally
+
+1. start the local replicate with `dfx start`
```
We should probably link to this doc https://github.com/dfinity/ic/tree/master/rs/rosetta-api/ledger_canister#deploying-locally to explain how to build and install the ledger.

MarioDfinity replied:
I've replaced the first two points with a link to that doc.

roman-kashitsyn commented on rust/tokens_transfer/src/lib.rs:
```diff
@@ -0,0 +1,67 @@
+        }
+    });
+    ic_ledger_types::transfer(ledger_canister_id, transfer_args).await
+        .map_err(|e| format!("ledger transfer error {:?}", e))?
```
```suggestion
        .map_err(|e| format!("failed to call ledger: {:?}", e))?
```

roman-kashitsyn commented on rust/tokens_transfer/src/tokens_transfer.did:
```diff
@@ -0,0 +1,32 @@
+
+type Memo = nat64;
+
+type TransferError = variant {
+    AccountError: text;
+    CallError: text;
+    LedgerError: text;
+};
+
+type TransferResult = variant {
+    Ok: Memo;
+    Err: TransferError;
+};
```
```suggestion
type TransferResult = variant {
    Ok: Memo;
    Err: text;
};
```
